### PR TITLE
Deferred torrent removal

### LINKED ISSUE #2907: Twistify the removal of torrents in the core
We should make the removal of torrents asynchronous. This is possible by using the `torrent_deleted_alert` (http://www.rasterbar.com/products/libtorrent/manual.html#torrent-deleted-alert) and return a deferred in the `remove_torrent` method in the `LibtorrentMgr` class.

This allows us to get rid of the ugly hard-coded timeouts when removing torrents (i.e. https://github.com/Tribler/tribler/blob/devel/Tribler/Core/APIImplementation/LaunchManyCore.py#L435).

### LINKED ISSUE #2969: Crash on attempting to change anonymity (to 0 hops, off)
- [x] I have read CONTRIBUTING.rst

- [x] I have tried with the latest pre-release version and I still can reproduce the issue.

##### Tribler version/branch+revision: #####
RC-2 2
##### Operating system and version: #####
Windows 10

##### Expected behavior: #####
Change happens, and Tribler continues as normally.

Possibly very important: Torrent was from a watch folder.

##### Actual behavior: #####
Tribler crashes as the change happens. On restart, the torrent remains as it was. 

##### Steps to reproduce the behavior: #####
Right click, change anonymity -> no anonymity. 

Likely caused by Tribler readding the torrent from the watch folder, before the one without anonymity is readded, causing the crash.   

##### Relevant log file output: #####
Forgot to copy-paste, but it got sent with the crash reporter.... sorry. Said duplicate torrent error though.

### PR DESCRIPTION
**Pull request by qstokkink**
Fixes #2907, fixes #2969.

### CODE REVIEW

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
@devos50 ready for review!

File: Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py [hunk 1]
```diff
@@ -1096,6 +1098,9 @@ def network_stop(self, removestate, removecontent):
 
             if removestate:
                 self.session.lm.remove_pstate(self.tdef.get_infohash())
+        if not out:
+            out = succeed(None)
+        return out
```
Comment (vandenheuvel) [inline Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py]: How about some sugar? `return out or succeed(None)`.

Comment (qstokkink) [inline Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py]: Good point, that would be a really sneaky way to get the line covered. 👍

File: Tribler/Test/Core/Libtorrent/test_libtorrent_mgr.py [hunk 2]
```diff
@@ -192,6 +192,26 @@ def test_add_torrent(self):
         self.assertEqual(self.ltmgr.add_torrent(None, {'ti': infohash}), mock_handle)
         self.assertRaises(DuplicateDownloadException, self.ltmgr.add_torrent, None, {'ti': infohash})
 
+    def test_remove_invalid_torrent(self):
+        """
+        Tests a successful removal status of torrents without a handle
+        """
+        self.ltmgr.initialize()
+        class MockDL(object):
```
Comment (devos50) [inline Tribler/Test/Core/Libtorrent/test_libtorrent_mgr.py]: Why don't use use the `MockObject` class for this to be consistent with the other test cases?

```
mock_dl = MockObject()
mock_dl.handle = None
```

The same goes for the `MockHandle` and `MockDL` classes you've introduced in the other test.

File: Tribler/Test/Core/test_session.py [hunk 3]
```diff
@@ -187,3 +188,28 @@ def verify_load_checkpoint_call():
...
+        dcfg.set_dest_dir(self.getDestDir())
+
+        # @deferred does not play nice with inlineCallbacks and blocking
+        @blocking_call_on_reactor_thread
+        @inlineCallbacks
+        def wait_for_dl_start():
+            yield self.session.start_download_from_tdef(torrent_def, dcfg=dcfg, hidden=True)
+        wait_for_dl_start()
+
+        return self.session.remove_download_by_id(torrent_def.infohash)
```
Comment (devos50) [inline Tribler/Test/Core/test_session.py]: Can you use the following (assuming the deferred returned from `start_download_from_tdef` fires with the infohash of the download)?

```
return self.session.start_download_from_tdef(torrent_def, dcfg=dcfg, hidden=True).addCallback(self.session.remove_download_by_id)
```

Comment (qstokkink) [inline Tribler/Test/Core/test_session.py]: 😊 Woops, `start_download_from_tdef` returns the `LibtorrentDownloadImpl`. This should've yielded on `download.get_handle()`. I can chain in a callback which forwards `handle.info_hash()` to `remove_download_by_id` 👍

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
@devos50 fixed 👍

**Pull request merged**